### ISSUE TITLE
`useChannel` is breaking with channel options

### ISSUE THREAD

**Issue by anthlasserre** [ably/ably-js#1570]
Here is my `useChannel` usage.

```typescript
useChannel(
    {
      channelName: channelName ?? "",
      options: {
        params: {
          rewind: "1",
          delta: "vcdiff",
        },
      },
    },
    onAblyMessageCallback,
  );
```

And an error is thrown on component mount.

```bash
Error: Channels.get() cannot be used to set channel options that would cause the channel to reattach. Please, use RealtimeChannel.setOptions() instead.
```

The issue is not happening every-time, it seems a bit random. But this is happening on component mount.

Here is my workaround until we find a solution.

```typescript
// 2. Subscribe to Ably messages
const { channels } = useAbly();

useEffect(() => {
  if (!channelName) {
    return;
  }
  const channel = channels.get(channelName);
  channel?.setOptions({
    params: {
      rewind: "1",
      delta: "vcdiff",
    },
  });
  channel?.subscribe(onAblyMessageCallback);

  return () => {
    channel?.unsubscribe();
  };
}, [channels]);
```

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-4131) by [Unito](https://www.unito.io)

**Comment by anthlasserre** [comment]
I needed to apply a custom logic to make it working...

```typescript
// 2. Subscribe to Ably messages
const { channels } = useAbly();

useEffect(() => {
  if (!channelName) {
    return;
  }
  const channel = channels.get(channelName);
  channel?.setOptions({
    params: {
      rewind: "1",
      delta: "vcdiff",
    },
  });
  channel?.subscribe(onAblyMessageCallback);

  return () => {
    channel?.unsubscribe();
  };
}, [channels]);
```

**Comment by ttypic** [comment]
Hi @anthlasserre,

Thank you for bringing this to our attention! The issue you're encountering is due to useChannel checking the options argument using shallow equals. If you modify your code as follows:


```
const options = useMemo(() => ({
    params: {
        rewind: "1",
        delta: "vcdiff",
    },
}), []);

useChannel(
    {
        channelName: channelName ?? "",
        options,
    },
    onAblyMessageCallback,
);

```

Everything should work as expected. I understand it might seem a bit complicated, and we are actively exploring ways to make this API more user-friendly. The primary concern we aim to address is avoiding situations where two channels have the same name but different options.

**Comment by anthlasserre** [comment]
Hey @ttypic! Still having the issue even with your memorization suggestion.

**Comment by ttypic** [comment]
Hey @anthlasserre, looks like your component that using `useChannel` hook unmounts from react tree at some point. Probably you can declare `options` outside component:

```js
const options = {
    params: {
        rewind: "1",
        delta: "vcdiff",
    },
}

const MyComponent = ({ channelName, onAblyMessageCallback }) => {
    useChannel({ channelName, options }, onAblyMessageCallback);
    return (...)
}
```

We are looking how to make it more user-friendly

**Comment by ttypic** [comment]
On second thought, `useChannel` checks params on shallow equals, so everything should work as expected. I think the problem is that you are using the same channel elsewhere in their app with different params

**Comment by codeKenta** [comment]
I'm also having this issue with the version `1.2.48`

In my case, it seems to be related to our app subscribing to channels from another Ably instance.

**Comment by ttypic** [comment]
Hey @codeKenta, @anthlasserre,

We released ably 1.2.49 with several bug fixes for `useChannel` hook. Please upgrade and let us know if it resolves your problems

**Comment by VeskeR** [comment]
Closing this as the fix for the issue was released in ably-js 1.2.49.
Also we greatly improved ably-js integration with react hooks in ably-js v2 release that also addresses issues like this one via utilizing a separate `ChannelProvider` component to set options for a channel, thus we recommend upgrading to ably-js v2 if possible.
